app/post_metadata.go

```diff
@@ -0,0 +1,418 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"image"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/dyatlov/go-opengraph/opengraph"
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/utils"
+	"github.com/mattermost/mattermost-server/utils/markdown"
+)
+
+const LINK_CACHE_SIZE = 10000
+const LINK_CACHE_DURATION = 3600
+
+var linkCache = utils.NewLru(LINK_CACHE_SIZE)
+
+func (a *App) InitPostMetadata() {
+	// Dump any cached links if the proxy settings have changed so image URLs can be updated
+	a.AddConfigListener(func(before, after *model.Config) {
+		if (before.ServiceSettings.ImageProxyType != after.ServiceSettings.ImageProxyType) ||
+			(before.ServiceSettings.ImageProxyURL != after.ServiceSettings.ImageProxyType) {
+			linkCache.Purge()
+		}
+	})
+}
+
+func (a *App) PreparePostListForClient(originalList *model.PostList) *model.PostList {
+	numPosts := len(originalList.Posts)
+
+	list := &model.PostList{
+		Posts: make(map[string]*model.Post, numPosts),
+		Order: originalList.Order, // Note that this uses the original Order array, so it isn't a deep copy
+	}
+
+	posts := make(chan *model.Post, numPosts)
+
+	for _, originalPost := range originalList.Posts {
+		go func(originalPost *model.Post) {
+			posts <- a.PreparePostForClient(originalPost)
+		}(originalPost)
+	}
+
+	for i := 0; i < numPosts; i++ {
+		post := <-posts
+		list.Posts[post.Id] = post
+	}
+
+	close(posts)
+
+	return list
+}
+
+func (a *App) PreparePostForClient(originalPost *model.Post) *model.Post {
+	post := originalPost.Clone()
```

Comment:
I was considering that, but it's lower priority than the other performance improvements I've been doing. Changing it here is easy enough, but I'd also need to change it in the code that adds the proxied image URLs to the post which may take some time because I don't understand that code as well